File: testing/mozbase/mozrunner/tests/test_wait.py
#!/usr/bin/env python
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.

from __future__ import absolute_import

import mozunit


def test_wait_while_running(runner):
    """Wait for the process while it is running"""
    runner.start()
    returncode = runner.wait(1)

    assert runner.is_running()
    assert returncode is None
    assert runner.returncode == returncode
    assert runner.process_handler is not None


def test_wait_after_process_finished(runner):
    """Bug 965714: wait() after stop should not raise an error"""
    runner.start()
    runner.process_handler.kill()

    returncode = runner.wait(1)

    assert returncode not in [None, 0]
    assert runner.process_handler is not None


if __name__ == '__main__':
    mozunit.main()

File: testing/marionette/test_wait.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

const {utils: Cu} = Components;

Cu.import("chrome://marionette/content/wait.js");

add_task(async function test_until_types() {
  for (let typ of [true, false, "foo", 42, [], {}]) {
    strictEqual(typ, await wait.until(resolve => resolve(typ)));
  }
});

add_task(async function test_until_timeoutElapse() {
  let nevals = 0;
  let start = new Date().getTime();
  await wait.until((resolve, reject) => {
    ++nevals;
    reject();
  });
  let end = new Date().getTime();
  greaterOrEqual((end - start), 2000);
  greaterOrEqual(nevals, 15);
});

add_task(async function test_until_rethrowError() {
  let nevals = 0;
  let err;
  try {
    await wait.until(() => {
      ++nevals;
      throw new Error();
    });
  } catch (e) {
    err = e;
  }
  equal(1, nevals);
  ok(err instanceof Error);
});

add_task(async function test_until_noTimeout() {
  // run at least once when timeout is 0
  let nevals = 0;
  let start = new Date().getTime();
  await wait.until((resolve, reject) => {
    ++nevals;
    reject();
  }, 0);
  let end = new Date().getTime();
  equal(1, nevals);
  less((end - start), 2000);
});

add_task(async function test_until_timeout() {
  let nevals = 0;
  let start = new Date().getTime();
  await wait.until((resolve, reject) => {
    ++nevals;
    reject();
  }, 100);
  let end = new Date().getTime();
  greater(nevals, 1);
  greaterOrEqual((end - start), 100);
});

add_task(async function test_until_interval() {
  let nevals = 0;
  await wait.until((resolve, reject) => {
    ++nevals;
    reject();
  }, 100, 100);
  equal(2, nevals);
});
